b2sdk/transfer/emerge/planner/planner.py

```diff
@@ -0,0 +1,493 @@
+        upload_intents = deque(upload_intents)
+        copy_intents = deque(copy_intents)
+        while True:
+            upload_intent = copy_intent = None
+            if upload_intents:
+                upload_intent, upload_end, _ = upload_intents[0]
+            if copy_intents:
+                copy_intent, copy_end, copy_protected = copy_intents[0]
+
+            if upload_intent is not None and copy_intent is not None:
+                if not copy_protected:
+                    yield_intent = upload_intent
+                else:
+                    yield_intent = copy_intent
+                start_offset = min(upload_end, copy_end)
+                yield yield_intent, start_offset
+                if start_offset >= upload_end:
+                    upload_intents.popleft()
+                if start_offset >= copy_end:
+                    copy_intents.popleft()
+            elif upload_intent is not None:
+                yield upload_intent, upload_end
+                upload_intents.popleft()
+            elif copy_intent is not None:
+                yield copy_intent, copy_end
+                copy_intents.popleft()
+            else:
+                return
+
+    def _validatation_iterator(self, write_intents):
+        last_offset = 0
+        for write_intent in write_intents:
+            if write_intent.length is None:
+                raise ValueError('Planner cannot support write intents of unknown length')
+            if write_intent.destination_offset < last_offset:
+                raise ValueError('Write intent stream have to be sorted by destination offset')
+            last_offset = write_intent.destination_offset
+            yield write_intent
+
+
+class IntentsState(object):
+    def __init__(self, protected_intent_length=0):
+        self.protected_intent_length = protected_intent_length
+        self._current_intent = None
+        self._next_intent = None
+        self._last_sent_offset = 0
+        self._incoming_offset = None
+        self._current_intent_end = None
+        self._next_intent_end = None
+
+    def add(self, incoming_intent):
+        if self._next_intent is None:
+            self._set_next_intent(incoming_intent)
+        elif incoming_intent.destination_end_offset > self._next_intent_end:
+            # here either incoming intent starts at the same position as next intent
+            # (and current intent is None in such case - it was just cleared in `state_update`
+            # or it was cleared some time ago - in previous iteratios) or we are in situation
+            # when current and next intent overlaps, and `last_sent_offset` is now set to
+            # incoming intent `destination_offset` - in both cases we want to choose
+            # intent which has larger `destination_end_offset`
+            self._set_next_intent(incoming_intent)
+
+    def state_update(self, last_sent_offset, incoming_offset):
```

Comment:
this function returns a 0 or 1-element iterator and the caller casts it to a list (but doesn't mutate the list later anyway, so it should probably be casted to a tuple). Why not just return a tuple from here?

Reply:
not true - it can return 0, 1 or 2 elements